src/persist-types/src/stats.rs

```diff
@@ -105,6 +114,83 @@ impl StructStats {
     }
 }
 
+// Aggregate statistics about a column of Json elements.
+//
+// Each element could be any of a JsonNull, a bool, a string, a numeric, a list,
+// or a map/object. The column might be a single type but could also be a
+// mixture of any subset of these types.
+#[derive(Default)]
+pub struct JsonStats {
+    /// The count of `Datum::JsonNull`s, or 0 if there were none.
+    pub json_nulls: usize,
+    /// The min and max bools, or None if there were none.
+    pub bools: Option<PrimitiveStats<bool>>,
+    /// The min and max strings, or None if there were none.
+    pub string: Option<PrimitiveStats<String>>,
+    /// The min and max numerics, or None if there were none.
+    ///
+    /// TODO(mfp): Storing this as an f64 is not correct.
+    pub numeric: Option<PrimitiveStats<f64>>,
+    /// The count of `Datum::List`s, or 0 if there were none.
+    ///
+    /// TODO: We could also do something for list indexes analogous to what we
+    /// do for map keys, but it initially seems much less likely that a user
+    /// would expect that to work with pushdown, so don't bother keeping the
+    /// stats until someone asks for it.
+    pub list: usize,
+    /// Recursive statistics about the set of keys present in any maps/objects
+    /// in the column, or None if there were no maps/objects.
+    pub map: BTreeMap<String, JsonStats>,
+}
+
+impl std::fmt::Debug for JsonStats {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let JsonStats {
+            json_nulls,
+            bools,
+            string,
+            numeric,
+            list,
+            map,
+        } = self;
+        let mut f = &mut f.debug_tuple("json");
+        if json_nulls > &0 {
+            f = f.field(json_nulls);
+        }
+        if let Some(bools) = bools {
+            f = f.field(bools);
+        }
+        if let Some(string) = string {
+            f = f.field(string);
+        }
+        if let Some(numeric) = numeric {
+            f = f.field(numeric);
+        }
+        if list > &0 {
+            f = f.field(list);
+        }
+        if !map.is_empty() {
+            f = f.field(map);
+        }
+        f.finish()
+    }
+}
+
+/// Statistics about a column of `Vec<u8>`.
+pub enum BytesStats {
+    Primitive(PrimitiveStats<Vec<u8>>),
+    Json(JsonStats),
```

Comment:
realized while do some followup work to use these stats that we need to do add a `BytesStats::OptJson<OptionalStats<JsonStats{..}>>` variant here instead of doing `OptionalStats<BytesStats::Json{..}>`

Reply:
Interesting - why's that?

Reply:
hmmm, the answer is that as written, all the `ProtoDatumStats` impls are mapping things to a columnar `Vec<u8>`, so for the nullable variants, we were encoding Datum::Null via ProtoDatum like any other datum. but now that I type that out, it seems dumb and we should make the nullable `ProtoDatumStats` impls use `Option<Vec<u8>>` like any other nullable field

Reply:
fixed! the math for `trait ProtoDatumStats` doesn't make as much sense for 2 impls as it did with 4 (plus we couldn't also do a blanket impl for a `trait OptProtoDatumStats` because rust rejects it as potentially conflicting impl), so backed that idea out